### PR TITLE
Added functionality to store datatype with values

### PR DESCRIPTION
**Pull request by s-shiva1995**
This can handle basic datatypes supported by firestore and reference now returns the path of references document from root collection with typeof as reference.

### CODE REVIEW

**Comment by yoiang** [comment]
Hey @s-shiva1995 , thanks so much for the contribution, it looks great!

I was wondering if there was a good alternative to using `typeof` on the same level as actual members inside of an object. It mixes metaphors, which is confusing, and also will conflict in the potential situation where an object contains an actual member named `typeof`. Potentially objects could use `value` similar to basic types where `value` would contain all of its members.

Also, this is more my personal taste, but I feel `typeof` more Javascript than Firestore, how do you feel about just calling the member `type`?

Finally I was wondering why you store members of an array as object rather than in an array?

**Comment by yoiang** [comment]
I've created a new branch starting with your PR: [s-shiva1995-patch-18](https://github.com/steadyequipment/node-firestore-backup/tree/s-shiva1995-patch-18)

It includes organization inspired by the functionality you added as well as Flow coverage of the new code. If you don't mind I'd like to use this as the continuation point!

**Comment by s-shiva1995** [comment]
Sure, well `typeof` or `type` both describe one and the same thing, but `type` keyword suits better.

About storing array also as object, I did this to simplify restoring code.  So that I can check which `type` of data is available in object and simply run the function to restore it accordingly.

**Comment by yoiang** [comment]
Gotcha! Ah you've already done work on the restore functionality, that's awesome 👍 

I sent you a PR with these changes, if it looks good updated your PR here and I'll merge it in!

**Comment by yoiang** [comment]
Ah thanks for catching those places I missed. I'd be good to have Flow coverage of that portion at some point. I'm going to spend a little trying to improve the [CODEBEAT score](https://codebeat.co/projects/github-com-steadyequipment-node-firestore-backup-master/pull_requests/93521) and then merge it!

File: lib/firestore.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    // String - string
+    if (isBoolean(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isBoolean(documentData[key]) })
+    } else if (isDate(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isDate(documentData[key]) })
+    } else if (isNumber(documentData[key])) {
+      documentDataToStore = Object.assign({}, documentDataToStore, { [key]: isNumber(documentData[key]) })
+    } else if (isArray(documentData[key])) {
+      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], { type: 'array' })
+      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], constructDocumentValue({}, Object.keys(documentData[key]), documentData[key]))
```
Comment (yoiang) [inline lib/firestore.js]: Is there a reason for assigning to `documentDataToStore[key]` twice? Would this be synonymous with 
```js
      documentDataToStore[key] = Object.assign({}, documentDataToStore[key], { type: 'array' }, constructDocumentValue({}, Object.keys(documentData[key]), documentData[key]))

```

or is there something that misses?

Comment (s-shiva1995) [reply lib/firestore.js]: You can do this in one statement also. This code is not optimised. So you can fix this type of code.
I will do optimisation once I am free.

**Comment by yoiang** [comment]
I'm going to merge this with the additional changes I pushed back at you unless there's anything else you'd like to get in!

**Pull request merged**